src/coreclr/jit/morph.cpp

```diff
@@ -13892,13 +13891,70 @@ PhaseStatus Compiler::fgMorphBlocks()
 
     // Morph all blocks.
     //
-    // Note morph can add blocks downstream from the current block,
-    // and alter (but not null out) the current block's bbNext;
-    // this iterator ensures they all get visited.
-    //
-    for (BasicBlock* block : Blocks())
+    if (!optLocalAssertionProp)
     {
-        fgMorphBlock(block);
+        // If we aren't optimizing, we just morph in normal bbNext order.
+        //
+        // Note morph can add blocks downstream from the current block,
+        // and alter (but not null out) the current block's bbNext;
+        // this iterator ensures they all get visited.
+        //
+        for (BasicBlock* block : Blocks())
+        {
+            fgMorphBlock(block);
+        }
+    }
+    else
+    {
+        // We are optimizing. Process in RPO.
+        //
+        fgRenumberBlocks();
+        EnsureBasicBlockEpoch();
+        fgComputeEnterBlocksSet();
+        fgDfsReversePostorder();
+
+        // Disallow general creation of new blocks or edges as it
+        // would invalidate RPO.
+        //
+        // Removal of edges, or altering dup counts, is OK.
+        //
+        INDEBUG(fgSafeBasicBlockCreation = false;);
+        INDEBUG(fgSafeFlowEdgeCreation = false;);
+
+        // Allow edge creation to genReturnBB (target of return merging)
+        // and the scratch block successor (target for tail call to loop).
+        // This will also disallow dataflow into these blocks.
+        //
+        if (genReturnBB != nullptr)
+        {
+            genReturnBB->bbFlags |= BBF_CAN_ADD_PRED;
+        }
+        if (fgFirstBBisScratch())
+        {
+            fgFirstBB->Next()->bbFlags |= BBF_CAN_ADD_PRED;
+        }
+
```

Comment:
It's unfortunate we need these special cases (and the new block flag), but I understand you tried to avoid it and this is a compromise.